paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc

```diff
@@ -336,10 +345,71 @@ proto::VarType::Type FuseOptimizerOpPass::GetTypeOfVar(
   return var_desc->GetType();
 }
 
+ir::Node *FuseOptimizerOpPass::CreateSyncTensorOpNode(
+    const std::vector<std::string> &in_args, const std::string &fused_out_arg,
+    BlockDesc *target_block, ir::Graph *result) const {
+  auto op_desc = target_block->AppendOp();
+  op_desc->SetType("sync_fused_tensor");
+  op_desc->SetInput("Input", in_args);
+  op_desc->SetInput("FusedInput", {fused_out_arg});
+  op_desc->SetOutput("FusedOutput", {fused_out_arg});
+  // NOTE: multi_devices_pass requires that every op should have a role.
+  op_desc->SetAttr(OpProtoAndCheckerMaker::OpRoleAttrName(),
+                   static_cast<int>(OpRole::kBackward));
+  return result->CreateOpNode(op_desc);
+}
+
+std::vector<ir::Node *> FuseOptimizerOpPass::GetVarNodesByName(
+    const std::vector<std::string> &var_names,
+    const std::unordered_map<std::string, std::vector<ir::Node *>> &vars_info)
+    const {
+  std::vector<ir::Node *> var_nodes;
+  for (auto &var_name : var_names) {
+    auto iter = vars_info.find(var_name);
+    PADDLE_ENFORCE_EQ(iter != vars_info.end(), true,
+                      "The Variable %s is not defined in program.", var_name);
+    for (auto var : iter->second) {
+      var_nodes.emplace_back(var);
+    }
+  }
+  return var_nodes;
+}
+
+void FuseOptimizerOpPass::InsertSyncTensorOpToGraph(
+    const std::vector<ir::Node *> &in_var_nodes, ir::Node *sync_tensor_op_node,
+    ir::Node *fused_opt_node, ir::Graph *result) const {
+  for (auto in_var_node : in_var_nodes) {
+    // set outpus
+    ir::Node *dep_var = result->CreateControlDepVar();
+    VLOG(6) << "Insert dep_var between " << sync_tensor_op_node->Op()->Type()
+            << " and " << fused_opt_node->Op()->Type();
+    sync_tensor_op_node->outputs.emplace_back(dep_var);
+    fused_opt_node->inputs.emplace_back(dep_var);
+    dep_var->inputs.emplace_back(sync_tensor_op_node);
+    dep_var->outputs.emplace_back(fused_opt_node);
```

Comment:
It seems that there not need so many `dep_var` between `sync_tensor_op_node ` and `fused_opt_node `.

Reply:
change to close this strategy in this scenario, new PR: https://github.com/PaddlePaddle/Paddle/pull/21005